bofire/strategies/doe/design.py

```diff
@@ -0,0 +1,255 @@
+
+from bofire.domain import Domain
+from bofire.domain.constraints import NChooseKConstraint
+from bofire.samplers import PolytopeSampler, apply_nchoosek
+from bofire.strategies.doe.jacobian import JacobianForLogdet
+from bofire.strategies.doe.utils import (
+    constraints_as_scipy_constraints,
+    get_formula_from_string,
+    metrics,
+    nchoosek_constraints_as_bounds,
+)
+
+
+def _domain_for_sampling(domain: Domain) -> Domain:
+    domain_for_sampling = deepcopy(domain)
+    # check if there are NChooseK constraints that must be ignored when sampling
+    if len(domain.cnstrs) > 0:
+        if any([isinstance(c, NChooseKConstraint) for c in domain.cnstrs]) and not all(
+            [isinstance(c, NChooseKConstraint) for c in domain.cnstrs]
+        ):
+            warnings.warn(
+                "Sampling of points fulfilling this problem's constraints is not implemented."
+            )
+
+            _constraints = []
+            for c in domain.cnstrs:
+                if not isinstance(c, NChooseKConstraint):
+                    _constraints.append(c)
+            domain_for_sampling = Domain(
+                input_features=domain.inputs,
+                output_features=domain.outputs,
+                constraints=_constraints,
+            )
+    return domain_for_sampling
+
+
+def logD(A: np.ndarray, delta: float = 1e-7) -> float:
+    """Computes the sum of the log of A.T @ A ignoring the smallest num_ignore_eigvals eigenvalues."""
+    X = A.T @ A + delta * np.eye(A.shape[1])
+    return np.linalg.slogdet(X)[1]
+
+
+def get_objective(
+    domain: Domain,
+    model_type: Union[str, Formula],
+    delta: float = 1e-7,
+) -> Callable:
+    """Returns a function that computes the objective value.
+
+    Args:
+        domain (Domain): A domain defining the DoE problem together with model_type.
+        model_type (str or Formula): A formula containing all model terms.
+        delta (float): Regularization parameter for information matrix. Default value is 1e-3.
+
+    Returns:
+        A function computing the objective -logD for a given input vector x
+
+    """
+    D = len(domain.inputs)
+    model_formula = get_formula_from_string(
+        model_type=model_type, rhs_only=True, domain=domain
+    )
+
+    # define objective function
+    def objective(x):
+        # evaluate model terms
+        A = pd.DataFrame(x.reshape(len(x) // D, D), columns=domain.inputs.get_keys())
+        A = model_formula.get_model_matrix(A)
+
+        # compute objective value
+        obj = -logD(A.to_numpy(), delta=delta)
+        return obj
+
+    return objective
+
+
+def find_local_max_ipopt(
+    domain: Domain,
+    model_type: Union[str, Formula],
+    n_experiments: Optional[int] = None,
+    tol: float = 0,
+    delta: float = 1e-7,
+    ipopt_options: Dict = {},
+    jacobian_building_block: Optional[Callable] = None,
+    sampling: Optional[np.ndarray] = None,
+    fixed_experiments: Optional[np.ndarray] = None,
```

Comment:
What do you think about modelling this as pandas dataframe, then we can also validate that it is formatted correctly by using `domain.validate_candidates(fixed_experiments, only_inputs=True)`. Furthermore, it fits more into the bofire way of using dataframes on all user exposed interfaces.